drivers/wifi/esp/esp_offload.c

```diff
@@ -552,14 +548,24 @@ static int esp_put(struct net_context *context)
 {
 	struct esp_socket *sock = context->offload_context;
 
-	esp_socket_workq_flush(sock);
+	esp_socket_workq_stop_and_flush(sock);
 
-	if (esp_socket_connected(sock)) {
+	if (esp_socket_flags_test_and_clear(sock, ESP_SOCK_CONNECTED)) {
 		esp_socket_close(sock);
 	}
 
+	k_mutex_lock(&sock->lock, K_FOREVER);
 	sock->connect_cb = NULL;
 	sock->recv_cb = NULL;
+	k_mutex_unlock(&sock->lock);
+
+	k_sem_reset(&sock->sem_free);
```

Comment:
This looks unnecessary. If you just want to know whether the ref count went to 0, just let `esp_socket_unref()` to return the value and use that directly.

Reply:
I actually want to be notified when refcount reaches 0. Call to `esp_socket_unref()` in this function might or might not be the last one. The reason is that there might be still some work in progress in esp_rx thread (parsing unsolicited AT command), so I want to wait until it finishes.

Reply:
Ok, could you add more description here how the `sem_free` is suppose to be working as it is not immediately clear to the reader what is going on here.

Reply:
Added comment to source code basically saying the same as comment in this thread.